Modules/Core/Common/include/itkBooleanStdVector.h

```diff
@@ -0,0 +1,43 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkBooleanStdVector_h
+#define itkBooleanStdVector_h
+
+#include "itkIntTypes.h"
+#include <vector>
+
+namespace itk
+{
+
+/** The element type of `BooleanStdVectorType`. Should be treated like a Boolean type. */
+using BooleanType = uint8_t;
+
+static_assert(sizeof(BooleanType) <= sizeof(bool), "itk::BooleanType should not be larger than the built-in C++ bool");
+static_assert(!BooleanType{ false }, "itk::BooleanType{ false } should be allowed as a conditional expression.");
+static_assert(BooleanType{ true }, "itk::BooleanType{ true } should be allowed as a conditional expression.");
+static_assert(bool{ BooleanType{ true } } == true && bool{ BooleanType{ false } } == false,
+              "itk::BooleanType should support lossless roundtip from a bool value.");
```

Comment:
Let's also assert about equality and logical and bitwise operators.  And if any of these fail, let's document that.
```c++
static_assert(BooleanType{ false } == !BooleanType{ true }, "itk::BooleanType is bad");
static_assert(BooleanType{ true } == !BooleanType{ false }, "itk::BooleanType is bad");
static_assert(BooleanType{ false } == ~BooleanType{ true }, "itk::BooleanType is bad");
static_assert(BooleanType{ true } == ~BooleanType{ false }, "itk::BooleanType is bad");
static_assert((BooleanType{ true } & BooleanType{ true }) == BooleanType{ true }, "itk::BooleanType is bad");
static_assert((BooleanType{ true } & BooleanType{ false }) == BooleanType{ false }, "itk::BooleanType is bad");
static_assert((BooleanType{ false } & BooleanType{ true }) == BooleanType{ false }, "itk::BooleanType is bad");
static_assert((BooleanType{ false } & BooleanType{ false }) == BooleanType{ false }, "itk::BooleanType is bad");
static_assert((BooleanType{ true } | BooleanType{ true }) == BooleanType{ true }, "itk::BooleanType is bad");
static_assert((BooleanType{ true } | BooleanType{ false }) == BooleanType{ true }, "itk::BooleanType is bad");
static_assert((BooleanType{ false } | BooleanType{ true }) == BooleanType{ true }, "itk::BooleanType is bad");
static_assert((BooleanType{ false } | BooleanType{ false }) == BooleanType{ false }, "itk::BooleanType is bad");
static_assert((BooleanType{ true } ^ BooleanType{ true }) == BooleanType{ false }, "itk::BooleanType is bad");
static_assert((BooleanType{ true } ^ BooleanType{ false }) == BooleanType{ true }, "itk::BooleanType is bad");
static_assert((BooleanType{ false } ^ BooleanType{ true }) == BooleanType{ true }, "itk::BooleanType is bad");
static_assert((BooleanType{ false } ^ BooleanType{ false }) == BooleanType{ false }, "itk::BooleanType is bad");
```

Reply:
Since you wrote these asserts, putting them in should be easy!

Reply:
@Leengit Thanks for the suggestion, but sorry, I find it a bit overkill to have so many asserts on such a simple typedef, `using BooleanType = uint8_t`.

Reply:
I get that.  Perhaps we could have these in a unit test file rather than in the include file that everyone must wait to compile.

Reply:
Personally, I think all these asserts are overkill.

Reply:
If all the asserts pass then they are overkill. :smile:  But if any fail then that means that someone who is hoping to get the `bool` behavior will instead get the `uint8_t` behavior.  At the very least, I'd want to document any different behavior in `itkBooleanStdVector.h` (even if the asserts live only in a unit test file).

Reply:
>someone who is hoping to get the bool behavior will instead get the uint8_t behavior

The whole point is to get the uint8_t behaviour because the bool behaviour is not thread safe.

Reply:
On my system, `uint8_t` has the same behavior as `bool` for all these equality, logical, and bitwise operators.  If this is the case for all platforms then the `static_asserts` are overkill.  

(Note that the tests I wrote above involving `~` are buggy; they insist that `uint8_t` do something that `bool` does not.)

Reply:
@Leengit Thanks, we could indeed add such a note. But of course, the proposed `BooleanStdVectorType` is only intended to be used for Boolean values, false and true. It's just not intended to be used to store arbitrary 8-bits integer values.

Reply:
@Leengit I hope I sufficiently addressed your suggestion for a note, by a [force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/dec13145e404b8b0448c575e9f16a260fbe19727..283d885183e788a00bc693f63f04750a696c4cef) which adds: 
```
 * \note BooleanStdVectorType is only intended to be used for the storage of Boolean values (true and false), not for
 * arbitrary integer values.
 */
```

Reply:
What I want to alert the reader to is ... sometimes people will write
```c++
bool flag0 = myInt0;
bool flag1 = myInt1;
```
when they mean
```c++
bool flag0 = myInt0 != 0;
bool flag1 = myInt1 != 0;
```
because the former is more concise and these two approaches have the exact same result.  In particular, one can follow either approach with
```c++
if (flag1 == flag2)
{
    // myInt0 and myInt1 were both zero or were both nonZero
};
```

However, if we put `unint8_t` in there instead of `bool` then the two approaches are no longer equivalent, and the `flag1 == flag2` test will not do what we are expecting under the concise approach when, say, `myInt0=10` and `myInt1=20`.

I think I've laid out my ideas here.  Whatever you think is best will work for me.

Reply:
> sometimes people will write
> 
> ```c++
> bool flag0 = myInt0;
> bool flag1 = myInt1;
> ```

But you won't write that, right? Wouldn't you do brace-initialization? As in:

    bool flag0 { myInt0 };
    bool flag1 { myInt1 };

Then the compiler would warn you against such a "narrowing conversion" 😸 

> I think I've laid out my ideas here. 

I agree that it's good to be aware of the fundamental differences between bool and uint8_t.

> Whatever you think is best will work for me.

Thanks @Leengit And if you do think so, please approve this pull request  😄

Reply:
You are correct that I would not use that concise approach.  If all programmers were like me, the world would be a different place.  (Yes, I know that that can be interpreted in multiple ways.)

I have reviewed and approved.  Thanks!